### PR TITLE
Exp/pretraining

### PR DESCRIPTION
**Pull request by sdascoli**

### CODE REVIEW

File: automatic_data_generation/data/base_dataset.py [hunk 1]
```diff
@@ -76,6 +77,9 @@ def __init__(self,
         label.build_vocab(train)
         intent.build_vocab(train)
 
+        self.input_type = input_type
+        self.embedding_dimension = embedding_dimension
+        self.embedding_type = embedding_type
```
Comment (choufractal) [inline automatic_data_generation/data/base_dataset.py]: `self.input_type` is already defined above. 
I would also set the other two at the beginning of the class.

File: automatic_data_generation/data/base_dataset.py [hunk 2]
```diff
@@ -271,3 +275,18 @@ def embed_unks(self, init="randn", num_special_toks=2):
...
+        # AND NOT
+        # vocab.itos = [w for w in vocab.itos if w not in loaded_itos] + loaded_itos 
+        # Order matters here ! We want the indices of the old dataset to be conserved...
+        vocab.stoi.update({w:i for (i,w) in enumerate(loaded_itos)})
+        
+    def update_vectors(self):
+        if self.embedding_type == 'glove':
+            emb_vectors = "glove.6B.{}d".format(self.embedding_dimension)
+            self.vocab.load_vectors(vectors=emb_vectors)
+        elif embedding_type == 'random':
```
Comment (choufractal) [inline automatic_data_generation/data/base_dataset.py]: missing a self here

File: automatic_data_generation/data/base_dataset.py [hunk 3]
```diff
@@ -271,3 +275,18 @@ def embed_unks(self, init="randn", num_special_toks=2):
...
+            self.vocab.vectors = torch.randn(len(self.vocab.itos), self.embedding_dimension)
```
Comment (choufractal) [inline automatic_data_generation/data/base_dataset.py]: I guess you need to fix a seed here too

File: automatic_data_generation/train_and_eval_cvae.py [hunk 4]
```diff
@@ -89,10 +97,10 @@ def train_and_eval_cvae(args):
     else:
         model = CVAE.from_folder(args.load_folder)
         LOGGER.info('Loaded model from %s' % args.load_folder)
-
-    # load embeddings
-    model.load_embedding(dataset.vectors)
-    LOGGER.debug('Model embedding', model.embedding)
+        model.n_classes = dataset.n_classes        
+        print(model.n_classes)
```
Comment (choufractal) [inline automatic_data_generation/train_and_eval_cvae.py]: use LOGGER.debut or LOGGER.info rather than print

**Review by choufractal — CHANGES_REQUESTED** [review]

File: automatic_data_generation/evaluation/metrics.py [hunk 5]
```diff
@@ -61,21 +61,25 @@ def calc_bleu(dataset, sentences, intents, type='utterance'):
...
+            bleu_scores['diversity'][intent] = np.mean(
+                [1-sentence_bleu(my_remove(candidates[intent],candidate), candidate, weights=[0.25, 0.25, 0.25, 0.25], smoothing_function=cc.method1)
+                 for candidate in candidates[intent]])
+            
+            # ORIGINAL DIVERSITY
+            bleu_scores['original_diversity'][intent] = np.mean(
+                [1-sentence_bleu(my_remove(references[intent],reference), reference, weights=[0.25, 0.25, 0.25, 0.25], smoothing_function=cc.method1)
+                 for reference in references[intent]])
+
+        except:
```
Comment (choufractal) [inline automatic_data_generation/evaluation/metrics.py:94]: You should avoid calling just `except` as it will catch any error and the necessarily the one you intended to in the first place

File: automatic_data_generation/evaluation/metrics.py [hunk 6]
```diff
@@ -137,20 +141,15 @@ def intent_classification(sentences, intents, train_path, input_type):
...
-                     / len([intent for intent in intents if intent != 'None'])) 
-
-    return accuracy
+    try:
+        intent_classifier = train_intent_classifier(train_path, input_type)
+        preds = intent_classifier.predict(sentences)
+        accuracy = float(sum([pred==intent for pred, intent in zip(preds,intents) if intent != 'None'])
+                     / len([intent for intent in intents if intent != 'None']))
+        return accuracy
+    except:
```
Comment (choufractal) [inline automatic_data_generation/evaluation/metrics.py:179]: same here

File: automatic_data_generation/data/base_dataset.py [hunk 7]
```diff
@@ -271,3 +281,49 @@ def embed_unks(self, init="randn", num_special_toks=2):
...
+        loaded_dict    = torch.load(str(folder / "vocab.pth"))
+        loaded_i2w     = loaded_dict['i2w']
+        loaded_i2int   = loaded_dict['i2int']
+        self.update_vocab(self.vocab, loaded_i2w)
+        self.update_vocab(self.intent.vocab, loaded_i2int)
+        self.update_vectors()
+
+        if self.input_type == 'delexicalised':
+            loaded_slotdic = loaded_dict['slotdic']
+            self.update_slotdic(loaded_slotdic)
```
Comment (choufractal) [inline automatic_data_generation/data/base_dataset.py:313]: it's not ideal because it won't work for any other class than `SnipsDataset`. To remedy this, I would create an abstract method in the Base class that raises a NotImplementedError

File: automatic_data_generation/data/handlers/snips_dataset.py [hunk 8]
```diff
@@ -74,8 +75,8 @@ def add_nones(sentences, none_folder, none_idx, none_size):
         return sentences
                                     
     def get_slotdic(self):
-        slotdic = {}
-        encountered_slot_values = {}
+        from collections import defaultdict
```
Comment (choufractal) [inline automatic_data_generation/data/handlers/snips_dataset.py]: imports should be on top

File: automatic_data_generation/data/base_dataset.py [hunk 9]
```diff
@@ -163,12 +167,18 @@ def build_data_files(self, dataset_folder, restrict_to_intent,
...
+        # trim_dataset : 
         trim_prefix = ''
         if dataset_size is not None:
-            # TODO: stratified shuffle split
             trim_prefix = '_{}'.format(dataset_size)
-            new_train = random.sample(new_train, dataset_size)
+            original_dataset_size = len(new_train)
+            keep_fraction = dataset_size/original_dataset_size
+            from sklearn.model_selection import StratifiedShuffleSplit
+            intents = [row[3] for row in new_train]
```
Comment (choufractal) [inline automatic_data_generation/data/base_dataset.py]: not correct if not SnipsDataset. You could fix that by addind `get_intents` as a child class method.

File: automatic_data_generation/train_and_eval_cvae.py [hunk 10]
```diff
@@ -157,8 +168,8 @@ def train_and_eval_cvae(args):
             logp,
             input_type='delexicalised'
         )
-        LOGGER.debug(*[{k: v} for (k, v) in run_dict[
-            'delexicalised_metrics'].items()], sep='\n')
+        LOGGER.info(*[{k: v} for (k, v) in run_dict[
+            'delexicalised_metrics'].items()])
```
Comment (choufractal) [inline automatic_data_generation/train_and_eval_cvae.py]: yields an error because some arguments cannot be converted to strings

File: automatic_data_generation/data/base_dataset.py [hunk 11]
```diff
@@ -163,12 +167,18 @@ def build_data_files(self, dataset_folder, restrict_to_intent,
...
-        # trim_dataset
+        # trim_dataset : 
         trim_prefix = ''
         if dataset_size is not None:
-            # TODO: stratified shuffle split
             trim_prefix = '_{}'.format(dataset_size)
-            new_train = random.sample(new_train, dataset_size)
+            original_dataset_size = len(new_train)
+            keep_fraction = dataset_size/original_dataset_size
+            from sklearn.model_selection import StratifiedShuffleSplit
```
Comment (choufractal) [inline automatic_data_generation/data/base_dataset.py]: imports should be on top on the module

**Review by choufractal — CHANGES_REQUESTED** [review]
